stdlib/public/core/Hashing.swift

```diff
@@ -147,51 +147,28 @@ func _mixInt(_ value: Int) -> Int {
 #endif
 }
 
-/// Given a hash value, returns an integer value within the given range that
-/// corresponds to a hash value.
+/// Given a hash value, returns an integer value between 0 and `upperBound`
+/// that corresponds to a hash value.
+///
+/// The `upperBound` must be positive and a power of 2.
 ///
 /// This function is superior to computing the remainder of `hashValue` by
 /// the range length.  Some types have bad hash functions; sometimes simple
 /// patterns in data sets create patterns in hash values and applying the
 /// remainder operation just throws away even more information and invites
-/// even more hash collisions.  This effect is especially bad if the length
-/// of the required range is a power of two -- applying the remainder
-/// operation just throws away high bits of the hash (which would not be
-/// a problem if the hash was known to be good).  This function mixes the
-/// bits in the hash value to compensate for such cases.
+/// even more hash collisions.  This effect is especially bad because the
+/// range is a power of two, which means to throws away high bits of the hash
+/// (which would not be a problem if the hash was known to be good). This
+/// function mixes the bits in the hash value to compensate for such cases.
 ///
 /// Of course, this function is a compressing function, and applying it to a
 /// hash value does not change anything fundamentally: collisions are still
 /// possible, and it does not prevent malicious users from constructing data
 /// sets that will exhibit pathological collisions.
 public // @testable
-func _squeezeHashValue(_ hashValue: Int, _ resultRange: Range<Int>) -> Int {
-  // Length of a Range<Int> does not fit into an Int, but fits into an UInt.
-  // An efficient way to compute the length is to rely on two's complement
-  // arithmetic.
-  let resultCardinality =
-    UInt(bitPattern: resultRange.upperBound &- resultRange.lowerBound)
-
-  // Calculate the result as `UInt` to handle the case when
-  // `resultCardinality >= Int.max`.
-  let unsignedResult =
-    _squeezeHashValue(hashValue, UInt(0)..<resultCardinality)
-
-  // We perform the unchecked arithmetic on `UInt` (instead of doing
-  // straightforward computations on `Int`) in order to handle the following
-  // tricky case: `startIndex` is negative, and `resultCardinality >= Int.max`.
-  // We cannot convert the latter to `Int`.
-  return
-    Int(bitPattern:
-      UInt(bitPattern: resultRange.lowerBound) &+ unsignedResult)
+func _squeezeHashValue(_ hashValue: Int, _ upperBound: Int) -> Int {
+  _sanityCheck(_isPowerOf2(upperBound))
+  let mixedHashValue = _mixInt(hashValue)
+  return mixedHashValue & (upperBound &- 1)
```

Comment:
Actually, this slipped my mind because I was expecting this code to exist already. I personally think its worth explaining this kind of strength reduction optimization to ensure the codebase can be maintained by non-experts. In this case just noting that this is an effecient idiom for x % bound for bounds that are a power of 2.